string_grouper/string_grouper.py

```diff
@@ -397,25 +407,21 @@ def _deduplicate(self) -> Union[pd.DataFrame, pd.Series]:
             ),
             shape=(n, n)
         )
-        raw_group_id_of_master_id = pd.DataFrame(
-            {
-                'raw_group_id': pd.Series(connected_components(csgraph=graph, directed=False)[1]),
-                'master_id': self._master.index.to_series()
-            }
-        )
-        first_master_id_in_group = raw_group_id_of_master_id.groupby('raw_group_id')['master_id']\
-            .first()\
-            .rename('new_group_id')\
-            .reset_index()
-        new_group_id_of_master_id = first_master_id_in_group\
-            .merge(raw_group_id_of_master_id, how='left', on='raw_group_id')\
-            .sort_values('master_id')\
-            .reset_index(drop=True)
-        output = self._master[new_group_id_of_master_id.new_group_id].reset_index(drop=True)
+        group_of_master_id = pd.Series(connected_components(csgraph=graph, directed=False)[1], name='raw_group_id')\
+            .reset_index()\
```

Comment:
Sure.  `pandas` sports many powerful hidden features. But unfortunately the downside is that code can become somewhat less readable.

Reply:
Now the new comments are a bit verbose but hopefully more clarifying.  But let me know if they are too verbose.

Reply:
Furthermore, while analyzing the structure of some sample groups, I accidentally discovered that originally the `StringGrouper._matches_list` DataFrame did not always turn out to be fully symmetric when `duplicates=None` &mdash; that is, some matched pairs of strings (A, B) in the list would also have their converse match (B, A) present in the list while others would have theirs absent.  This led to erroneous centroid determinations.   I've fixed that now.

On another note, the group-structure analysis showed me that the centroid is indeed a natural choice for group-representative.

Reply:
> Now the new comments are a bit verbose but hopefully more clarifying. But let me know if they are too verbose.

Thanks! Yes this is much better! I don't think its to verbose (although maybe this line is a bit: # convert to pandas Series:)

> Furthermore, while analyzing the structure of some sample groups, I accidentally discovered that originally the StringGrouper._matches_list DataFrame did not always turn out to be fully symmetric when duplicates=None — that is, some matched pairs of strings (A, B) in the list would also have their converse match (B, A) present in the list while others would have theirs absent. This led to erroneous centroid determinations. I've fixed that now.

Hmm, this is strange, i thought it should always be symmetric. Good find! Could you add a unit test that should fail if this function isn't present? I would like to know how it happens. Maybe it happens because of the `max_n_matches` config? 

For example if max_n_matches = 1, and there is a match A --> B, the other match (B --> A) would drop. 

> On another note, the group-structure analysis showed me that the centroid is indeed a natural choice for group-representative.

Great!

Reply:
> Could you add a unit test that should fail if this function isn't present? I would like to know how it happens. 

Done.

> Maybe it happens because of the max_n_matches config?
> For example if max_n_matches = 1, and there is a match A --> B, the other match (B --> A) would drop.

That is quite likely the case, and is consistent with the fact that I found this bug while working on a large dataset (the `seg__edgar_company_info.csv` sample file)